kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -136,51 +139,81 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches() {
     // Helper Methods //
     ////////////////////
 
-    private Tuple2<Protocol, Metadata> loadTableProtocolAndMetadata() {
+    private Tuple2<Protocol, Metadata> loadTableProtocolAndMetadata(
+            Optional<SnapshotHint> snapshotHint) {
         Protocol protocol = null;
         Metadata metadata = null;
 
-        try (CloseableIterator<Tuple2<FileDataReadResult, Boolean>> reverseIter =
+        try (CloseableIterator<ActionIterElem> reverseIter =
                  new ActionsIterator(
                      tableClient,
                      logSegment.allLogFilesReversed(),
                      PROTOCOL_METADATA_READ_SCHEMA)) {
             while (reverseIter.hasNext()) {
-                final ColumnarBatch columnarBatch = reverseIter.next()._1.getData();
+                final ActionIterElem nextElem = reverseIter.next();
+                final ColumnarBatch columnarBatch = nextElem.getFileDataReadResult().getData();
+                final long version = nextElem.getVersion();
 
                 assert(columnarBatch.getSchema().equals(PROTOCOL_METADATA_READ_SCHEMA));
 
                 if (protocol == null) {
-                    final ColumnVector protocolVector = columnarBatch.getColumnVector(0);
+                    if (snapshotHint.isPresent() && version <= snapshotHint.get().getVersion()) {
```

Comment:
This should be `==` -- `<=` isn't safe in the general case where the hint could be earlier than the earliest log segment version (e.g. hint at v101, checkpoint at v105). Granted, in such cases we _should_ always get a P&M from the checkpoint, but if somehow the checkpoint were incorrect/incomplete, I don't think we'd want the hint to hide that fact?

Reply:
Agreed!

Reply:
Agreed it should be `==` ... but just FYI in your example, `snapshotHint.get().getVersion() = 101`, `version=105`

then `version <= snapshotHint.get().getVersion()` evaluates to `false` so we wouldn't use the hint

Reply:
Right... I had it backward. The only way the `==` vs `<=` matters is if we had a gap in the version number sequence, but the listing logic already checked for that.